rmgpy/molecule/kekulize.pyx

```diff
@@ -0,0 +1,290 @@
+3. Determine if any bonds in the ring are already defined (not benzene bonds).
+4. For the remaining bonds, determine whether or not they can be double bonds.
+5. If a clear determination cannot be made, make heuristic based assumption.
+6. Continue until all bonds in the ring are determined.
+7. Continue until all rings in the molecule are determined.
+"""
+
+from .molecule cimport Atom, Bond, Molecule
+from .element import PeriodicSystem
+
+
+cpdef kekulize(Molecule mol):
+    """Kekulize an aromatic molecule."""
+    cdef list ring, rings, aromaticRings, resolvedRings
+    cdef set endoBonds, exoBonds
+    cdef Atom atom1, atom2
+    cdef Bond bond
+    cdef bint aromatic, successful
+    cdef int itercount, maxiter
+    cdef AromaticRing aromaticRing
+
+    # Get all potentially aromatic rings
+    rings = mol.getAllCyclesOfSize(6)
+
+    # Identify aromatic rings and categorize endocyclic and exocyclic bonds for each ring
+    aromaticRings = []
+    for ring in rings:
+        endoBonds = set()
+        exoBonds = set()
+        aromatic = True
+        for atom1 in ring:
+            for atom2, bond in atom1.bonds.iteritems():
+                if atom2 in ring:
+                    if abs(round(bond.order) - bond.order) < 1e-9:
+                        aromatic = False
+                        break
+                    endoBonds.add(bond)
+                else:
+                    exoBonds.add(bond)
+            if not aromatic:
+                break
+        if aromatic:
+            aromaticRings.append(AromaticRing(atoms=ring, endoBonds=endoBonds, exoBonds=exoBonds))
+
+    resolvedRings = []
+    itercount = 0
+    maxiter = 2 * len(aromaticRings)
+    while aromaticRings and itercount < maxiter:
+        # Update and sort the remaining rings
+        prioritizeRings(aromaticRings)
+        # Take the next ring off the stack
+        aromaticRing = aromaticRings.pop()
+        # Try to kekulize this ring
+        successful = aromaticRing.kekulize()
+        if successful:
+            resolvedRings.append(aromaticRing)
+        else:
+            aromaticRings.append(aromaticRing)
+        itercount += 1
+
+    mol.updateAtomTypes(logSpecies=False)
+
+cdef list prioritizeRings(list aromaticList):
+    """Update list of AromaticRing objects, then sort by DOF."""
+    cdef AromaticRing item, x
+    for item in aromaticList:
+        item.update()
+    return aromaticList.sort(key=lambda x: (x.endoDOF, x.exoDOF), reverse=True)
+
+cdef list prioritizeBonds(list aromaticList):
+    """Update list of Aromatic Bond objects, then sort by DOF."""
+    cdef AromaticBond item, x
+    for item in aromaticList:
+        item.update()
+    return aromaticList.sort(key=lambda x: (x.endoDOF, x.exoDOF), reverse=True)
+
+cdef class AromaticRing:
+    """Helper class containing information about a single aromatic ring in a molecule."""
+    cdef list atoms
+    cdef set endoBonds, exoBonds
+    cdef public int endoDOF, exoDOF
```

Comment:
Again, endo refers to bonds that comprise the ring, and exo refers to bonds that are connected to the ring, but not a part of it. DOF stands for degrees of freedom, and refers to the number of bonds that do not have a set bond order yet.